PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,317 @@
+            }
+            self.originalAppUserId = originalAppUserIdString
+
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions = subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(withSubscriptionsData: nonSubscriptions, dateFormatter: DateFormatter.dateFormatter)
+
+            let latestNonSubscriptionTransactionsByProductId = [String: Any](uniqueKeysWithValues: nonSubscriptions.map { productId, transactionsArray in
+                (productId, transactionsArray.last)
+            })
+
+            self.allPurchases = latestNonSubscriptionTransactionsByProductId.merging(subscriptions) { (current, _) in current }
+
+            self.expirationDatesByProduct = PurchaserInfo.parseExpirationDate(expirationDates: subscriptions)
+            self.purchaseDatesByProduct = PurchaserInfo.parsePurchaseDate(purchaseDates: allPurchases)
+        }
+    }
+
+}
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set<String>(dates.keys.filter({
+            if let nonNullDate = dates[$0] as? Date {
+                return isAfterReferenceDate(date: nonNullDate)
+            }
+            return true
+        }))
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: Any) -> URL? {
+        if let urlString = url as? String {
+            return URL.init(string: urlString)
+        }
+        return nil
+    }
+
+    class func parseDate(date: Any, withDateFormatter: DateFormatter) -> Date? {
+        if let dateString = date as? String {
+            return DateFormatter.dateFormatter.date(from: dateString)
+        }
+        return nil
+    }
+
+    class func parseExpirationDate(expirationDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: expirationDates, label: "expires_date") }
+
+    class func parsePurchaseDate(purchaseDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: purchaseDates, label: "purchase_date") }
+
+    class func parseDatesIn(dates: [String: Any], label: String) -> [String: Date?] {
+        return dates.mapValues {
+            if let dateString = ($0 as? [String: String] ?? [String: String]())[label] {
+                return DateFormatter.dateFormatter.date(from: dateString)
+            } else {
+                return nil
+            }
+        }
+    }
+
+}
+
+private extension DateFormatter {
+
+    static let dateFormatter: DateFormatter = {
```

Comment:
this syntax reads a little weird to me to be honest: 
```Swift
DateFormatter.dateFormatter.foo
```

perhaps this could just be a variable on the class? so it's just `dateFormatter.foo`?

Reply:
agreed it looked weird, was following the `use private extensions of basic types to define constants` in swift sample file, but i suppose this isn't quite the case you were referring to... will update! 👍

Reply:
@aboedo made it a static member. still requires `Self.`, but i think it's a little more clear. let me know your thoughts